File: frontend/src/lib/doctors.ts
import { post, request } from "@/lib/api";

export type Title = "Dr" | "Prof" | "Mr" | "Ms" | "Mrs";
export type DoctorStatus = "active" | "inactive";

export const TITLES: Title[] = ["Dr", "Prof", "Mr", "Ms", "Mrs"];

/** Monday first, the way a rota is read and written. */
export const DAYS = [
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
  "Sunday",
] as const;

export const SHORT_DAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"] as const;

export type ScheduleBlock = {
  id: string;
  day_of_week: number;
  start_time: string;
  end_time: string;
  break_start: string | null;
  break_end: string | null;
  slot_duration_minutes: number | null;
};

/** A block being edited, before it has been saved and given an id. */
export type BlockDraft = Omit<ScheduleBlock, "id"> & { key: string };

export type Leave = {
  id: string;
  starts_on: string;
  ends_on: string;
  start_time: string | null;
  end_time: string | null;
  reason: string | null;
  created_at: string;
};

export type DoctorSummary = {
  id: string;
  display_name: string;
  full_name: string;
  title: Title;
  speciality: string | null;
  qualifications: string | null;
  room: string | null;
  status: DoctorStatus;
  consultation_fee: string;
  fee_from_clinic: boolean;
  slot_duration_minutes: number;
  working_days: number;
  has_account: boolean;
  created_at: string;
};

export type Doctor = DoctorSummary & {
  first_name: string;
  last_name: string;
  registration_number: string | null;
  years_of_experience: number | null;
  phone: string | null;
  email: string | null;
  languages: string[] | null;
  bio: string | null;
  follow_up_fee: string;
  follow_up_fee_from_clinic: boolean;
  own_consultation_fee: string | null;
  own_follow_up_fee: string | null;
  own_slot_duration_minutes: number | null;
  user_id: string | null;
  account_name: string | null;
  deactivated_at: string | null;
  schedule: ScheduleBlock[];
  leaves: Leave[];
};

export type DoctorDraft = {
  title: Title;
  first_name: string;
  last_name: string;
  speciality?: string | null;
  qualifications?: string | null;
  registration_number?: string | null;
  years_of_experience?: number | null;
  phone?: string | null;
  email?: string | null;
  room?: string | null;
  languages?: string[] | null;
  bio?: string | null;
  consultation_fee?: string | null;
  follow_up_fee?: string | null;
  slot_duration_minutes?: number | null;
  user_id?: string | null;
};

export type DoctorPage = {
  items: DoctorSummary[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
};

export type Slot = { start_time: string; end_time: string };

export type Availability = {
  date: string;
  day_of_week: number;
  day_name: string;
  working: boolean;
  reason: string | null;
  slot_duration_minutes: number;
  slots: Slot[];
};

export type DoctorQuery = {
  q?: string;
  speciality?: string;
  status?: "active" | "inactive" | "all";
  page?: number;
  per_page?: number;
};

export const PAGE_SIZE = 25;

export function listDoctors({
  q = "",
  speciality = "",
  status = "active",
  page = 1,
  per_page = PAGE_SIZE,
}: DoctorQuery = {}): Promise<DoctorPage> {
  const params = new URLSearchParams({
    status,
    page: String(page),
    per_page: String(per_page),
  });
  if (q.trim()) params.set("q", q.trim());
  if (speciality.trim()) params.set("speciality", speciality.trim());
  return request<DoctorPage>(`/doctors?${params}`);
}

export const getDoctor = (id: string) => request<Doctor>(`/doctors/${id}`);

export const listSpecialities = (status: DoctorQuery["status"] = "active") =>
  request<string[]>(`/doctors/specialities?status=${status}`);

export const addDoctor = (draft: DoctorDraft) => post<Doctor>("/doctors", draft);

export const saveDoctor = (id: string, changes: Partial<DoctorDraft>) =>
  request<Doctor>(`/doctors/${id}`, { method: "PATCH", body: JSON.stringify(changes) });

export const deactivateDoctor = (id: string) => post<Doctor>(`/doctors/${id}/deactivate`, {});

export const restoreDoctor = (id: string) => post<Doctor>(`/doctors/${id}/restore`, {});

export const getSchedule = (id: string) => request<ScheduleBlock[]>(`/doctors/${id}/schedule`);

export const saveSchedule = (id: string, blocks: Omit<ScheduleBlock, "id">[]) =>
  request<ScheduleBlock[]>(`/doctors/${id}/schedule`, {
    method: "PUT",
    body: JSON.stringify({ blocks }),
  });

export const getAvailability = (id: string, date: string) =>
  request<Availability>(`/doctors/${id}/availability?date=${date}`);

export const recordLeave = (
  id: string,
  leave: {
    starts_on: string;
    ends_on?: string | null;
    start_time?: string | null;
    end_time?: string | null;
    reason?: string | null;
  },
) => post<Leave>(`/doctors/${id}/leaves`, leave);

export const cancelLeave = (doctorId: string, leaveId: string) =>
  request<{ removed: boolean }>(`/doctors/${doctorId}/leaves/${leaveId}`, {
    method: "DELETE",
  });

/** Initials for the avatar, from whichever names exist. */
export function initials(name: string): string {
  const parts = name.trim().split(/\s+/).filter(Boolean);
  const first = parts.at(0);
  if (!first) return "?";
  if (parts.length === 1) return first.slice(0, 2).toUpperCase();
  return `${first.charAt(0)}${parts.at(-1)?.charAt(0) ?? ""}`.toUpperCase();
}

/** "09:00:00" as "9:00 am", which is how a clinic writes its hours up. */
export function readableTime(value: string | null): string {
  if (!value) return "";
  const [hours = NaN, minutes = 0] = value.split(":").map(Number);
  if (Number.isNaN(hours)) return value;
  const suffix = hours < 12 ? "am" : "pm";
  const twelve = hours % 12 === 0 ? 12 : hours % 12;
  return `${twelve}:${String(minutes).padStart(2, "0")} ${suffix}`;
}

/** How long a slot runs, in minutes. */
export function slotMinutes(slot: Slot): number {
  const at = (value: string) => {
    const [hours = 0, minutes = 0] = value.split(":").map(Number);
    return hours * 60 + minutes;
  };
  return at(slot.end_time) - at(slot.start_time);
}

/**
 * How the free times read at the bottom of the list.
 *
 * A length is only named when every slot is that long. A clinic that runs
 * its evening at ten minutes a head and its morning at thirty has neither
 * number as the answer, and picking one of them is worse than picking none.
 */
export function countSlots(slots: Slot[]): string {
  const many = slots.length === 1 ? "1 appointment" : `${slots.length} appointments`;
  const lengths = new Set(slots.map(slotMinutes));
  const only = lengths.size === 1 ? [...lengths][0] : null;
  return only ? `${many} of ${only} minutes.` : `${many}.`;
}

/** What a time input needs, from what the API sends back. */
export const forInput = (value: string | null): string => (value ? value.slice(0, 5) : "");

export function readableDate(value: string): string {
  return new Date(`${value}T00:00:00`).toLocaleDateString("en-IN", {
    day: "numeric",
    month: "short",
    year: "numeric",
  });
}

export function describeLeave(leave: Leave): string {
  const span =
    leave.starts_on === leave.ends_on
      ? readableDate(leave.starts_on)
      : `${readableDate(leave.starts_on)} – ${readableDate(leave.ends_on)}`;
  if (!leave.start_time) return span;
  return `${span}, ${readableTime(leave.start_time)} – ${readableTime(leave.end_time)}`;
}

/** A line describing a doctor at a glance under their name. */
export function describe(doctor: DoctorSummary): string {
  return [doctor.speciality, doctor.qualifications].filter(Boolean).join(" · ");
}

/** The days of the week a rota touches, in order, for the list row. */
export function workingDays(blocks: ScheduleBlock[]): number[] {
  return [...new Set(blocks.map((item) => item.day_of_week))].sort((a, b) => a - b);
}

/** Today at the browser, as the date input and the API both want it. */
export function todayISO(): string {
  const now = new Date();
  const local = new Date(now.getTime() - now.getTimezoneOffset() * 60_000);
  return local.toISOString().slice(0, 10);
}

export function shiftDate(value: string, days: number): string {
  const day = new Date(`${value}T00:00:00`);
  day.setDate(day.getDate() + days);
  const local = new Date(day.getTime() - day.getTimezoneOffset() * 60_000);
  return local.toISOString().slice(0, 10);
}

/**
 * Strips what a form produces but the API should not read as a value.
 *
 * A fee has to survive this as null rather than as "0": null means "follow
 * the clinic" and zero means "this doctor sees people for nothing".
 */
export function cleaned(draft: DoctorDraft): DoctorDraft {
  const blank = (value: unknown) => typeof value === "string" && value.trim() === "";
  const tidied = Object.fromEntries(
    Object.entries(draft).map(([key, value]) => [key, blank(value) ? null : value]),
  ) as DoctorDraft;

  const languages = (tidied.languages ?? []).map((one) => one.trim()).filter(Boolean);
  tidied.languages = languages.length ? languages : null;
  return tidied;
}

File: backend/app/api/v1/endpoints/doctors.py
"""Doctors, their weekly hours, and who is free when.

Reading is open to everyone who books, checks in or bills; changing a
profile or a rota is not. Every read goes through the scoped repository, so
another clinic's doctor is not found rather than forbidden.
"""

from __future__ import annotations

import datetime as dt
import math
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import Caller, DbSession, Trail, current_tenant, requires
from app.models import Organization
from app.models.doctor import ACTIVE
from app.repositories.doctors import (
    DoctorRepository,
    LeaveRepository,
    Listed,
    ScheduleRepository,
    names_of,
)
from app.schemas.doctor import (
    Availability,
    DoctorCreate,
    DoctorOut,
    DoctorPage,
    DoctorSummary,
    DoctorUpdate,
    LeaveOut,
    LeaveWrite,
    Removed,
    ScheduleBlockOut,
    ScheduleWrite,
)
from app.services import doctors as service
from app.services import events

router = APIRouter(tags=["doctors"])

_DAYS = ("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")


async def _clinic(session: AsyncSession, organization_id: uuid.UUID) -> Organization:
    clinic = await session.get(Organization, organization_id)
    if clinic is None:
        raise service.DoctorNotFound
    return clinic


def _summary(row: Listed, clinic: Organization) -> DoctorSummary:
    doctor = row.doctor
    fee, from_clinic = service.effective_fee(
        doctor.consultation_fee, clinic, "consultation_fee"
    )
    return DoctorSummary(
        id=doctor.id,
        display_name=doctor.display_name,
        full_name=doctor.full_name,
        title=doctor.title,
        speciality=doctor.speciality,
        qualifications=doctor.qualifications,
        room=doctor.room,
        status=doctor.status,
        consultation_fee=fee,
        fee_from_clinic=from_clinic,
        slot_duration_minutes=service.effective_slot_minutes(doctor, clinic),
        working_days=row.working_days,
        has_account=doctor.user_id is not None,
        created_at=doctor.created_at,
    )


async def _read(
    session: AsyncSession, organization_id: uuid.UUID, doctor_id: uuid.UUID
) -> DoctorOut:
    row = await DoctorRepository(session, organization_id).with_working_days(doctor_id)
    if row is None:
        raise service.DoctorNotFound

    doctor = row.doctor
    clinic = await _clinic(session, organization_id)

    schedule = await ScheduleRepository(session, organization_id).for_doctor(doctor_id)
    leaves = await LeaveRepository(session, organization_id).for_doctor(
        doctor_id, ending_after=service.clinic_today(clinic)
    )
    names = await names_of(session, {doctor.user_id} if doctor.user_id else set())

    follow_up, follow_up_from_clinic = service.effective_fee(
        doctor.follow_up_fee, clinic, "follow_up_fee"
    )

    return DoctorOut(
        **_summary(row, clinic).model_dump(),
        first_name=doctor.first_name,
        last_name=doctor.last_name,
        registration_number=doctor.registration_number,
        years_of_experience=doctor.years_of_experience,
        phone=doctor.phone,
        email=doctor.email,
        languages=doctor.languages,
        bio=doctor.bio,
        follow_up_fee=follow_up,
        follow_up_fee_from_clinic=follow_up_from_clinic,
        # Carried separately from the figure that applies, so the edit form
        # can show a blank box that means "follow the clinic" rather than
        # writing the clinic's number into the profile the moment it opens.
        own_consultation_fee=(
            None if doctor.consultation_fee is None else f"{doctor.consultation_fee:.2f}"
        ),
        own_follow_up_fee=(
            None if doctor.follow_up_fee is None else f"{doctor.follow_up_fee:.2f}"
        ),
        own_slot_duration_minutes=doctor.slot_duration_minutes,
        user_id=doctor.user_id,
        account_name=names.get(doctor.user_id) if doctor.user_id else None,
        deactivated_at=doctor.deactivated_at,
        schedule=[ScheduleBlockOut.model_validate(block) for block in schedule],
        leaves=[LeaveOut.model_validate(leave) for leave in leaves],
    )


@router.get("/doctors")
async def list_doctors(
    session: DbSession,
    query: str = Query(default="", max_length=120, alias="q"),
    speciality: str = Query(default="", max_length=80),
    status: str = Query(default=ACTIVE, pattern="^(active|inactive|all)$"),
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=25, ge=1, le=100),
    _: Caller = Depends(requires("doctor:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorPage:
    clinic = await _clinic(session, organization_id)
    found, total = await DoctorRepository(session, organization_id).search(
        query=query,
        speciality=speciality,
        status=None if status == "all" else status,
        limit=per_page,
        offset=(page - 1) * per_page,
    )
    return DoctorPage(
        items=[_summary(row, clinic) for row in found],
        total=total,
        page=page,
        per_page=per_page,
        pages=max(math.ceil(total / per_page), 1),
    )


@router.get("/doctors/specialities")
async def list_specialities(
    session: DbSession,
    status: str = Query(default=ACTIVE, pattern="^(active|inactive|all)$"),
    _: Caller = Depends(requires("doctor:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> list[str]:
    """What this clinic actually offers, rather than a fixed list every
    clinic has to pick the wrong answer from.

    Takes the same status as the list so the two agree; a filter that can
    only ever come back empty is worse than no filter.
    """
    return await DoctorRepository(session, organization_id).specialities(
        None if status == "all" else status
    )


@router.post("/doctors", status_code=201)
async def add_doctor(
    session: DbSession,
    body: DoctorCreate,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorOut:
    doctor = await service.add(session, organization_id=organization_id, body=body)
    found = await _read(session, organization_id, doctor.id)
    await trail("doctor.added", "doctor", found.id, found.display_name)
    return found


@router.get("/doctors/{doctor_id}")
async def read_doctor(
    session: DbSession,
    doctor_id: uuid.UUID,
    _: Caller = Depends(requires("doctor:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorOut:
    return await _read(session, organization_id, doctor_id)


@router.patch("/doctors/{doctor_id}")
async def update_doctor(
    session: DbSession,
    doctor_id: uuid.UUID,
    body: DoctorUpdate,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorOut:
    fields = body.model_dump(exclude_unset=True)
    current = await DoctorRepository(session, organization_id).get(doctor_id)
    before = events.snapshot(current, fields) if current else {}
    await service.update(
        session, organization_id=organization_id, doctor_id=doctor_id, body=body
    )
    found = await _read(session, organization_id, doctor_id)
    changed = await DoctorRepository(session, organization_id).get(doctor_id)
    moved = events.difference(before, events.snapshot(changed, fields)) if changed else None
    if moved:
        await trail("doctor.changed", "doctor", found.id, found.display_name, moved)
    return found


@router.post("/doctors/{doctor_id}/deactivate")
async def deactivate_doctor(
    session: DbSession,
    doctor_id: uuid.UUID,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorOut:
    """Takes a doctor off the lists without losing their rota or anything
    signed in their name."""
    await service.set_active(
        session, organization_id=organization_id, doctor_id=doctor_id, active=False
    )
    found = await _read(session, organization_id, doctor_id)
    await trail("doctor.deactivated", "doctor", found.id, found.display_name)
    return found


@router.post("/doctors/{doctor_id}/restore")
async def restore_doctor(
    session: DbSession,
    doctor_id: uuid.UUID,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> DoctorOut:
    await service.set_active(
        session, organization_id=organization_id, doctor_id=doctor_id, active=True
    )
    found = await _read(session, organization_id, doctor_id)
    await trail("doctor.restored", "doctor", found.id, found.display_name)
    return found


@router.get("/doctors/{doctor_id}/schedule")
async def read_schedule(
    session: DbSession,
    doctor_id: uuid.UUID,
    _: Caller = Depends(requires("doctor:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> list[ScheduleBlockOut]:
    await service.fetch(session, organization_id=organization_id, doctor_id=doctor_id)
    blocks = await ScheduleRepository(session, organization_id).for_doctor(doctor_id)
    return [ScheduleBlockOut.model_validate(block) for block in blocks]


@router.put("/doctors/{doctor_id}/schedule")
async def write_schedule(
    session: DbSession,
    doctor_id: uuid.UUID,
    body: ScheduleWrite,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> list[ScheduleBlockOut]:
    """The whole week at once. Sending an empty list clears it."""
    blocks = await service.replace_schedule(
        session, organization_id=organization_id, doctor_id=doctor_id, body=body
    )
    written = [ScheduleBlockOut.model_validate(block) for block in blocks]
    found = await _read(session, organization_id, doctor_id)
    week = [
        f"{_DAYS[block.day_of_week]} {block.start_time:%H:%M}-{block.end_time:%H:%M}"
        for block in written
    ]
    await trail(
        "doctor.hours_set",
        "doctor",
        found.id,
        found.display_name,
        {"week": ", ".join(week) or "No hours"},
    )
    return written


@router.get("/doctors/{doctor_id}/availability")
async def read_availability(
    session: DbSession,
    doctor_id: uuid.UUID,
    date: dt.date | None = Query(default=None),
    _: Caller = Depends(requires("doctor:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> Availability:
    clinic = await _clinic(session, organization_id)
    day = date or service.clinic_today(clinic)
    found = await service.availability(
        session,
        organization_id=organization_id,
        clinic=clinic,
        doctor_id=doctor_id,
        day=day,
    )
    return Availability.model_validate(found)


@router.post("/doctors/{doctor_id}/leaves", status_code=201)
async def record_leave(
    session: DbSession,
    doctor_id: uuid.UUID,
    body: LeaveWrite,
    trail: Trail,
    caller: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> LeaveOut:
    leave = await service.add_leave(
        session,
        organization_id=organization_id,
        doctor_id=doctor_id,
        actor_id=caller.user_id,
        body=body,
    )
    written = LeaveOut.model_validate(leave)
    found = await _read(session, organization_id, doctor_id)
    await trail(
        "doctor.leave_added",
        "doctor",
        found.id,
        found.display_name,
        {"from": written.starts_on, "to": written.ends_on, "reason": written.reason},
    )
    return written


@router.delete("/doctors/{doctor_id}/leaves/{leave_id}")
async def cancel_leave(
    session: DbSession,
    doctor_id: uuid.UUID,
    leave_id: uuid.UUID,
    trail: Trail,
    _: Caller = Depends(requires("doctor:manage")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> Removed:
    found = await _read(session, organization_id, doctor_id)
    leave = await LeaveRepository(session, organization_id).get(leave_id)
    held = {"from": leave.starts_on, "to": leave.ends_on} if leave else None
    await service.remove_leave(
        session,
        organization_id=organization_id,
        doctor_id=doctor_id,
        leave_id=leave_id,
    )
    await trail("doctor.leave_removed", "doctor", found.id, found.display_name, held)
    return Removed()
